pkg/gadgets/trace/tcpdrop/tracer/tracer.go

```diff
@@ -0,0 +1,208 @@
+		str = strings.TrimPrefix(str, "SKB_")
+
+		t.dropReasons[int(v.Value)] = str
+	}
+
+	return nil
+}
+
+func (t *Tracer) install() error {
+	err := t.loadDropReasons()
+	if err != nil {
+		return err
+	}
+
+	spec, err := loadTcpdrop()
+	if err != nil {
+		return fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	gadgets.FixBpfKtimeGetBootNs(spec.Programs)
+
+	opts := ebpf.CollectionOptions{}
+
+	if err := spec.LoadAndAssign(&t.objs, &opts); err != nil {
+		return fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	t.kfreeSkbLink, err = link.Tracepoint("skb", "kfree_skb", t.objs.IgTcpdrop, nil)
+	if err != nil {
+		return fmt.Errorf("opening tracepoint kfree_skb: %w", err)
+	}
+
+	reader, err := perf.NewReader(t.objs.tcpdropMaps.Events, gadgets.PerfBufferPages*os.Getpagesize())
+	if err != nil {
+		return fmt.Errorf("creating perf ring buffer: %w", err)
+	}
+	t.reader = reader
+
+	return nil
+}
+
+func (t *Tracer) run() {
+	for {
+		record, err := t.reader.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				// nothing to do, we're done
+				return
+			}
+
+			msg := fmt.Sprintf("reading perf ring buffer: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			return
+		}
+
+		if record.LostSamples > 0 {
+			msg := fmt.Sprintf("lost %d samples", record.LostSamples)
+			t.eventCallback(types.Base(eventtypes.Warn(msg)))
+			continue
+		}
+
+		bpfEvent := (*tcpdropEvent)(unsafe.Pointer(&record.RawSample[0]))
+
+		reason, err := t.lookupDropReason(int(bpfEvent.Reason))
+		if err != nil {
+			msg := fmt.Sprintf("looking up drop reason: %s", err)
+			t.eventCallback(types.Base(eventtypes.Err(msg)))
+			continue
+		}
+
+		event := types.Event{
+			Event: eventtypes.Event{
+				Type:      eventtypes.NORMAL,
+				Timestamp: gadgets.WallTimeFromBootTime(bpfEvent.Timestamp),
+			},
+			// mount namespace not actually related to the container that created the connection
+			// Dropping the mount namespace ID for now
+			// TODO: add a way to get the mount namespace ID of the container that created the connection
+			// WithMountNsID: eventtypes.WithMountNsID{MountNsID: bpfEvent.MntnsId},
+			Pid:      bpfEvent.Pid,
+			Comm:     gadgets.FromCString(bpfEvent.Comm[:]),
+			Dport:    gadgets.Htons(bpfEvent.Dport),
+			Sport:    gadgets.Htons(bpfEvent.Sport),
+			State:    tcpbits.TCPState(bpfEvent.State),
+			Tcpflags: tcpbits.TCPFlags(bpfEvent.Tcpflags),
+			Reason:   reason,
+		}
+
+		if bpfEvent.Af == unix.AF_INET {
+			event.IPVersion = 4
+		} else if bpfEvent.Af == unix.AF_INET6 {
+			event.IPVersion = 6
+		}
+
+		event.Saddr = gadgets.IPStringFromBytes(bpfEvent.Saddr, event.IPVersion)
+		event.Daddr = gadgets.IPStringFromBytes(bpfEvent.Daddr, event.IPVersion)
+
+		t.eventCallback(&event)
+	}
+}
+
+func (t *Tracer) lookupDropReason(reason int) (string, error) {
+	if ret, ok := t.dropReasons[reason]; ok {
+		return ret, nil
+	}
+	return "", fmt.Errorf("unknown drop reason: %d", reason)
+}
+
+// --- Registry changes
```

Comment:
This comment was added to the other gadgets to show which parts were added for the refactoring. It can be removed here. Maybe it would also make sense to move the exported functions to the top in the order `NewInstance()`, `Run()`, `SetEventHandler()` to make it easier to read/find, but we also don't have that for the other gadgets, yet, so feel free to ignore :)